utils/litemap/src/store/vec_impl.rs

```diff
@@ -104,38 +104,116 @@ impl<K, V> StoreMut<K, V> for Vec<(K, V)> {
     {
         self.retain(|(k, v)| predicate(k, v))
     }
-}
 
-impl<K: Ord, V> StoreFromIterable<K, V> for Vec<(K, V)> {
-    fn lm_sort_from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
+    /// Extends this store with items from an iterator.
+    ///
+    /// It uses a three-pass approach to avoid any potential quadratic costs.
```

Comment:
issue: this implementation increases the capacity of the map by at least m/2, *and* also uses m auxiliary space in the worst case

I think this algorithm has equivalent runtime complexity, but only uses m storage:

1. append all the new items, O(m)
2. sort the appended items, O(m log(m)), Ω(m)
3. deduplicate the appended items against themselves, O(m)
4. if the first appended item is less than the last existing item
    - deduplicate the appended items against existing items, O(m log(n))
    - quicksort the whole thing, O((n+m) log(n+m))

The happy path here is still Ω(m), and the worst case is still O((n+m) log(n+m)), but less space is used.

Reply:
Maybe I'm missing something, but the deduplication in place is O(n²) due to the memmove calls.

Reply:
You don't do a sequence of O(n) removals. Instead you iterate with two pointers, one pointing at the next element, and one pointing at the end of the deduplicated sequence (both starting at 0). If you encounter a duplicated element, you increment the next-element pointer (which will always be at least the other pointer), if you encounter a unique element you replace what's currently at the end of the deduplicated sequence (everything between the pointers is trash). When the next-element pointer reaches the end of the list, you remove the trash.

This requires some thought for non-Copy types, but the algorithm is definitely linear.

Reply:
I had to modify partition_dedup to keep the last duplicated entry instead of the first, but otherwise, it's similar to what you suggested. The only drawback is that it always reserves the lower bound length of the incoming iterator as it doesn't make sense to only reserve half.

The new benchmark numbers are below. Without dups it's slightly faster than my original solution. With dups it's roughly the same.

```
litemap/extend_rand/large
                        time:   [66.882 ms 67.852 ms 69.096 ms]
litemap/extend_rand_dups/large
                        time:   [174.22 ms 176.15 ms 178.28 ms]

Reply:
I settled on a simplified version that is within +- 8% of the longer version depending on the benchmark.